merging development into main.

echo2334 opened the pull request:
all basic functionality of robot is here and working as intended. features include L2, L3, L4, ramp, manual effector, mecanum drive, strafe drift correction, do nothing auto, move past start line auto, and preliminary code for limelight. limelight features will be implemented more in the advanced code.

echo2334 committed:
Created Elevator.java

There is errors surrounding the motors. It is currently in simulator mode.

echo2334 committed:
Added all of auto

working on gyro

Commit:
moving?

Commit:
nescio

FRC8334Students committed:
created the effector file and class.

echo2334 committed:
stuff

FRC8334Students committed:
Working on limelight and april tag code

It's almost ready

impossiblerocket committed:
auto and gyro working

-figure out how to print the degrees of gyro(taylor's being only semi helpful but will ask again on saturday)
-TEST turndegrees action
-TEST drivefortime action
-gyro can be pulled to other branches and added onto
-idk what i was doing in robot
-added the correct library for gyro

FRC8334Students committed:
Created distance monitor

echo2334 committed:
Simulation of Elevator

The simulation can run now!!!!! I have the basic code base for the elevator. Need to test the base in the simulation. The file is currently a mess.

echo2334 committed:
added basic framework for how the end effector will work.

echo2334 committed:
gyro works. navx got bought out so the changed the names of things.

echo2334 committed:
Elise was working on the elevator

echo2334 committed:
started on the logic for scoring control. the idea is to have it be as easy as Aeode was to operate

echo2334 committed:
added even more logic. specifically the logic for when to switch cases on some of them, and added operator controller to teleop to control one of the states.

echo2334 committed:
gyro works in auto!

-HOW do we initialize mecanum in robot?
-once we get that everything should work
-took away the default code stuff from robot(i have no idea where it came from)

echo2334 committed:
whoever was last on this appears to not have committed, so i am committing now

echo2334 committed:
Merge branch 'Elevator' of https://github.com/Team8334/2025REDOBuildSeason into Elevator

echo2334 committed:
fixed some duplicate logic in scoring control, created EncoderValues.java, removed a taylor testing print statement, and added the eject feature to operator control, used to score or to drop coral

echo2334 committed:
Working on stuff

echo2334 committed:
small thing

BismuthTime committed:
adding in cases for elevator presets that will be provided by Elise. the massive case statement might be condensed later with refinement, but this should be a decent start.

echo2334 committed:
Try to do presets

I had gotten the simulation to work. I have been trying to intergrade presets with a controller. Currently the code is not responding to the controller.

echo2334 committed:
Can see apriltag+identify it+identify where it is

echo2334 committed:
Driving in auto works!

-turning does not work, gyro is getting all degrees but is getting two speed inputs when tries to turn?
-mecanum code is definitely the newest here
-DRIVING forward is actually driving backward, do not know why yet

echo2334 committed:
Turning in auto works!

-gyro has a reset function now and does reset to 0 after a 360 degree turn
-turning works!

FRC8334Students committed:
Aiden got the first draft of a more complex drive base done

FRC8334Students committed:
found and issue in mecanum where motors were being assigned to the wrong ports.

FRC8334Students committed:
copied the gyro from rhea's auto to start using it to fix mecanum stuff.

FRC8334Students committed:
Scoring logic should be all ready aside from elevator stuff. added Flight Controller (which i totally didn't name after myself). flight controller should be better for preset operation.

echo2334 committed:
Trying for encoder values

Not reciving the encoder values. set future plans to put case statement for presets.

echo2334 committed:
organized missions

-The controller is probably inverted, forward is actually -1
     This would explain why auto is going backwards
-added some comments to the missions so that we know what is happening with them

echo2334 committed:
Finished new drive system

echo2334 committed:
tun and backward should work

echo2334 committed:
stuff

echo2334 committed:
Finished

Can find, lock onto, and drive towards targeted apriltag. Needs an auto action to actually use the code though.

echo2334 committed:
fixed issues in scoring control, switched back to a normal controller, commented out some print statements, and added prints to test safety features

echo2334 committed:
Turn and forward in auto both work!

- forward is NEGATIVE SPEED
- need to merge with aiden's encoder code to stop the drift
- then moveAcrossLineMission should be working well but can technically work without this
- multiple actions at once do work
- need effactor code to do L1 auto
- need elevator code to do L4 auto
- limelight auto would probably be useful

echo2334 committed:
fixed scoring logic

echo2334 committed:
Merge branch 'LimelightandAprilTag' into Auto

echo2334 committed:
working on limelight actions

echo2334 committed:
Effector action set up

- effector action is set up but needs elevator code and the actual scoring code (merging branches) to be tested
- L1 scoring missions are set up but need effector action and field measurements
- started looking at drive for distance action, need position estimator
- added position estimator, but need encoder values from mecanum, might be able to do without position estimator and pull straight from mecanum code?
- DO NOT USE DRIVE FOR DISTANCE OR POSITION ESTIMATOR FOR SCRIMMAGE

echo2334 committed:
adding stuff

echo2334 committed:
aahhh

echo2334 committed:
Merge branch 'Auto' of https://github.com/Team8334/2025REDOBuildSeason into Auto

echo2334 committed:
yay names

echo2334 committed:
good pid

echo2334 committed:
Encoder

Encoder gets value does not change

echo2334 committed:
config in prep to test

Elised9 committed:
Encoder

Done research. Needs to be tested.

echo2334 committed:
Merge branch 'effector' into Development

echo2334 committed:
Merge branch 'SpeedControl' into Development

echo2334 committed:
deletetions

deleted position estimator and drive for distance action because won't be testable by scrimmage

echo2334 committed:
idk

echo2334 committed:
Auto guesses

echo2334 committed:
auto things

echo2334 committed:
specifying robots

echo2334 committed:
Finished lock on action front

Front lock on action should be ready

echo2334 committed:
Merge branch 'Auto' of https://github.com/Team8334/2025REDOBuildSeason into Auto

echo2334 committed:
Motor

Encoder works. Needs to take Motor and use encoder with it.

Elised9 committed:
Motors with encoders

NOT TESTED. Hopefully, this works.

Elised9 committed:
Trying

So, I made a few differences. I don't know if stuff would work the way I'm thinking.

echo2334 committed:
Merge branch 'Elevator' into Development

echo2334 committed:
stuff

echo2334 committed:
Merge branch 'nescio-quid-clamare-hoc' into Development

echo2334 committed:
scrimmage new effector code

echo2334 committed:
yay

echo2334 committed:
effectors should work now

echo2334 committed:
Effector works!

echo2334 committed:
encoder values now go above one and should work

echo2334 committed:
changing seconds of moving forward

FRC8334Students committed:
Working speed control code

echo2334 committed:
added set voltage and stuff

echo2334 committed:
auto here same as auto in auto branch

echo2334 committed:
removed set goal from stop so it actually stops.

echo2334 committed:
Merge branch 'Elevator' into temporaryDevelopment

echo2334 committed:
added elevator logic to scoring control. finally!!!

FRC8334Students committed:
laser can works

FRC8334Students committed:
Update ScoringControl.java

echo2334 committed:
elevator L2 and ramp and passive works!!!

echo2334 committed:
Merge branch 'temporaryDevelopment' into Auto

echo2334 committed:
all presets working as intended.

echo2334 committed:
merged temporary dev into dev.

echo2334 committed:
drive base encoders

- trying to get encoder distance
    / need to make sure encoders on robot and if so port
      numbers
    / looking at last years code for position estimation
- idk what the error means, ask jonathon for help
- END GOAL: use the drive for distance action which is a lot easier to use on the go instead of drive for time action

echo2334 committed:
all elevator functions working as intended.

echo2334 committed:
Merge branch 'Development' into Auto

echo2334 committed:
Merge branch 'SpeedControlFinal' into Development

echo2334 committed:
Have an elevator action

IS NOT TESTED
- need to check logic
- got rid of position estimation because wasn't working

echo2334 committed:
Moving with Elevator up

Done all the code. NEEDS TO BE TESTED!!!!!

echo2334 committed:
Effector action and Elevator action need to be tested

- do not expect to work

FRC8334Students committed:
cleaned up every class. all basic functionality of robot working as intended. will be merged into main.

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Actions/DriveForTimeAction.java:
```diff
@@ -0,0 +1,51 @@
+    @Override
+    public void done() {
+        timer.stop();
+        mDrive.driveWithSpeed(0, 0,0);
```
```suggestion
        mDrive.driveWithSpeed(0, 0, 0);
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Actions/DriveForTimeAction.java:
```diff
@@ -0,0 +1,51 @@
+
+/* This action moves the robot forward for a certain number
+ * of seconds at a certain speed
+ * negative is FORWARD, positive is backward
```
Is this actually true?

echo2334 replied:
yes, negative is forward and positive is backwards. drive for time working as intended aside from that.

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Actions/EffectorAction.java:
```diff
@@ -0,0 +1,47 @@
+package frc.robot.Auto.Actions;
+
+//imports important things
```
Most helpful comment

echo2334 replied:
yeah yeah

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Actions/EffectorAction.java:
Have we actually tested this entire file?

echo2334 replied:
i merged rhea's better code into dev, so it is be good now.

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Actions/EffectorAction.java:
```diff
@@ -0,0 +1,47 @@
+
+    @Override
+    public void update(){
+        score.EffectorStateProcessing(); // NEED TO ADD DOUBLE SECONDS AND SCORING STATE TO SCORINGCONTROL?
```
You have this comment about a thing you need to do, Is this done already or do we need to track a task to do this?

echo2334 replied:
rhea's new code does.

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Actions/EffectorAction.java:
```diff
@@ -0,0 +1,47 @@
+    public void done(){
+        timer.stop();
+        score.EffectorStateProcessing();
+        System.out.println("Scoring in L1 done");
```
I would keep system.out.println lines out, will slow down the robot
```suggestion
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Actions/FrontFollowAction.java:
Don't add an empty file?

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Actions/FrontLockOnAction.java:
Remove empty file?

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Actions/MoveElevatorAction.java:
```diff
@@ -0,0 +1,43 @@
+import edu.wpi.first.wpilibj.Timer;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+
+//import frc.robot.Subsystem.Elevator;//CHANGE ACCORDING TO REAL NAME
```
```suggestion
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Actions/MoveElevatorAction.java:
```diff
@@ -0,0 +1,43 @@
+    private double speed;
+    Timer timer;
+
+    public MoveElevatorAction (double seconds, double speed){// add another one for stage?
```
Comment here still relevant?

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Actions/TurnDegreesAction.java:
```diff
@@ -0,0 +1,81 @@
+import frc.robot.Subsystem.Mecanum;
+import frc.robot.Devices.Gyro;
+
+// imported math things
```
Very helpful

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Actions/TurnDegreesAction.java:
```diff
@@ -0,0 +1,81 @@
+ * This action turns the robot a certain number of degrees
+ * for a certain number of seconds
+ * negative is left and positive is right
+ * May immplement pid at some point
```
```suggestion
 * May implement pid at some point
```
Is this done? This part of the comment isn't helpful

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Actions/TurnDegreesAction.java:
```diff
@@ -0,0 +1,81 @@
+        SmartDashboard.putNumber("targetDegrees ", targetDegrees);
+        SmartDashboard.putNumber("desiredDegrees ", desiredDegrees);
+        currentDegrees = gyro.getAngleDegrees(); // gets the current degrees
+        System.out.println("currentDegrees: " + currentDegrees);
```
```suggestion
```

dalittlet replied:
And the line below it

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Actions/TurnDegreesAction.java:
```diff
@@ -0,0 +1,81 @@
+        currentDegrees = gyro.getAngleDegrees(); // gets the current degrees
+        System.out.println("currentDegrees: " + currentDegrees);
+        System.out.println("targetDegrees: " + targetDegrees);
+        SmartDashboard.putNumber("currentDegreese ", currentDegrees);
```
```suggestion
        SmartDashboard.putNumber("currentDegrees ", currentDegrees);
```

echo2334 replied:
sigh.

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Actions/TurnDegreesAction.java:
```diff
@@ -0,0 +1,81 @@
+
+    @Override
+    public boolean isFinished(){
+        if (targetDegrees == currentDegrees || timer.get() >= seconds){
```
Here me out, this entire function is just
```suggestion
        return targetDegrees == currentDegrees || timer.get() >= seconds;
```

and remove lower lines

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Actions/TurnDegreesAction.java:
```diff
@@ -0,0 +1,81 @@
+    public void done(){
+        SmartDashboard.putString( "Current Action", "TurnDegreesAction Ended");
+        SmartDashboard.putNumber("turnDegreesAction/speed" , turn);
+        //currentDegrees = targetDegrees;
```
```suggestion
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/AutoMissionChooser.java:
```diff
@@ -0,0 +1,141 @@
+
+public class AutoMissionChooser {
+    enum DesiredMission {
+        exampleMission,
```
Maybe add comments on what these missions are? Like what is example mission?

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/AutoMissionChooser.java:
```diff
@@ -0,0 +1,141 @@
+        missionChooser.addOption("Leave Community", DesiredMission.MoveAcrossLineMission);
+        missionChooser.addOption("Scoring L1", DesiredMission.ScoringL1Mission);
+        missionChooser.addOption("Scoring L4", DesiredMission.ScoringL4Mission);
+        missionChooser.addOption("Testing", DesiredMission.Testing);
```
What do this do?

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/AutoMissionChooser.java:
```diff
@@ -0,0 +1,141 @@
+        switch (mission) {
+            // do nothing mission
+            case doNothing:
+                return Optional.of(new DoNothingMission());
```
What do these Optionals do?

echo2334 replied:
its a state machine to determine what auto to run based on the smart dashboard selection.

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/AutoMissionEndedException.java:
```diff
@@ -0,0 +1,8 @@
+ * Exception thrown when an auto mission ends unexpectedly
+ */
+public class AutoMissionEndedException extends Exception {
+    private static final long serialVersionUID = 1411131586291540143L;
```
What is this for?

echo2334 replied:
after significant research into this, i don't know how it works. it assists exception handling in auto. it works, so.

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/AutoMissionExecutor.java:
```diff
@@ -0,0 +1,77 @@
+    private Thread mThread = null;
+
+    public void setAutoMission(MissionBase new_auto_Mission) {
+        mAutoMission = new_auto_Mission;
```
Is there a reason we are switching from everything else being camelCase to under_score?

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/AutoMissionExecutor.java:
```diff
@@ -0,0 +1,77 @@
+
+    public void start() {
+        if (mThread != null) {
+            //System.out.println("mThread is not null");
```
```suggestion
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/AutoMissionExecutor.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    public boolean isInterrupted() {
+        if (mAutoMission == null) {
```
Could do this whole function as 
```suggestion
        return mAutoMission == null ? false :  mAutoMission.getIsInterrupted();
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Missions/BlueScoreL1.java:
```diff
@@ -0,0 +1,26 @@
+    @Override
+    protected void routine() throws AutoMissionEndedException {
+       
+        runAction(new WaitAction(AutoMissionChooser.delay)); // MAY NOT BE NEEDED
```
When are we determining if this is needed?

echo2334 replied:
that comment was misleading. fixed now

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Missions/BlueScoreL1.java:
```diff
@@ -0,0 +1,26 @@
+       
+        runAction(new WaitAction(AutoMissionChooser.delay)); // MAY NOT BE NEEDED
+        runAction(new DriveForTimeAction(-0.5,  4));
+        runAction(new TurnDegreesAction(65, 1));// add 5 degrees because michalangelo has consistently com up 5 degrees short of the needed turn
```
A radio comes up 5 degrees short of the turn?

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Missions/BlueScoreL4.java:
```diff
@@ -0,0 +1,27 @@
+    @Override
+    protected void routine() throws AutoMissionEndedException {
+       
+        runAction(new WaitAction(AutoMissionChooser.delay)); // MAY NOT BE NEEDED
```
Same comments as BlueScoreL1

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Missions/MissionBase.java:
```diff
@@ -0,0 +1,105 @@
+ * An abstract class that is the basis of the robot's autonomous routines. This is implemented in auto missions (which are routines that do actions).
+ */
+
+    protected final double mUpdateRate = 1.0 / 50.0;
```
What is this for and what's the math behind it?

echo2334 replied:
we need to talk with aiden about this one. might try to get rhea to pester him about it on saturday (the 8th).

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Missions/MissionBase.java:
```diff
@@ -0,0 +1,105 @@
+    protected abstract void routine() throws AutoMissionEndedException;
+
+    public void setStartPose() {
+        //Drive.getInstance().setHeading(mStartPose.getRotation()); //if one day track position
```
```suggestion
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Missions/MoveAcrossLineMission.java:
```diff
@@ -0,0 +1,23 @@
+    @Override
+    protected void routine() throws AutoMissionEndedException {
+       
+        //runAction(new WaitAction(AutoMissionChooser.delay)); // MAY OR MAY NOT BE NEEDED
```
```suggestion
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Missions/Testing.java:
```diff
@@ -0,0 +1,23 @@
+        runAction(new TurnDegreesAction(-110, 3));
+        runAction(new DriveForTimeAction (0.2, 2));;
+        runAction(new WaitAction(1));
+        //runAction(new TurnDegreesAction(180, 3));
```
```suggestion
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Data/PortMap.java:
```diff
@@ -2,10 +2,24 @@
-    }
 
+        public static final float MOVING_THRESHOLD = (float)(0.0);
+        public static final float MAX_HEIGHT_FOR_DRIVING = (float)(10000000000000000000.0);
```
Just why

echo2334 replied:
idk

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Devices/Controller.java:
```diff
@@ -32,6 +31,26 @@ else if (this.getRawButtonPressed(button.value))
+        if (!debounceButtons.containsKey(button.value)) {
+            debounceButtons.put(button.value, false);
+        }
+        if (this.getRawButton(button.value) && debounceButtons.get(button.value))
```
Just a sanity check, we don't want this an if else if with the statement above correct?

echo2334 replied:
i think these are going after two separate things. it works right now, so.

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Devices/Limelight.java:
```diff
@@ -0,0 +1,152 @@
+    // distance from the center of the Limelight lens to the floor NEED VALUE FOR ROBOT
+    protected double limelightLensHeightMeters = 0.2; 
+
+	private String alliance = "Red";
```
Misaligned tab

echo2334 replied:
it is resolved on my side.

dalittlet replied:
Wrong, diff says otherwise

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Devices/Limelight.java:
```diff
@@ -0,0 +1,152 @@
+    public String findTagName(){
+        switch (getId())
+        {
+            case 12, 13 -> {
```
Want your case statements to match other case statements in other files?

echo2334 replied:
its using lambda statements, and i don't want to change it as it might effect limelight.

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Devices/Limelight.java:
```diff
@@ -0,0 +1,152 @@
+    }
+    
+    public boolean isOperational(){
+        if(limelightName == null){
```
```suggestion
        return limelightName != null;
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Devices/ModifiedEncoders.java:
```diff
@@ -0,0 +1,144 @@
+    }
+
+    public Boolean isConnected(){
+        if(dutyCycleEncoder != null){
```
```suggestion
        return dutyCycleEncoder != null ? dutyCycleEncoder.isConnected() : false;
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Devices/ModifiedEncoders.java:
```diff
@@ -0,0 +1,144 @@
+    }
+    
+    public int getFrequency(){
+        if(dutyCycleEncoder != null){
```
Could make a ternary expression (see line 36)

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Devices/ModifiedEncoders.java:
```diff
@@ -0,0 +1,144 @@
+    }
+    
+    public double get(){
+        if(dutyCycleEncoder != null){
```
ternary expression

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Devices/ModifiedEncoders.java:
```diff
@@ -0,0 +1,144 @@
+        if((saveEncoder - currentValue) >= 0.5){
+            cycle += 1;
+        }
+        if((saveEncoder - currentValue) < -0.5){
```
```suggestion
       else {
```

echo2334 replied:
i don't think so. in the first one we are checking its relation to 0.5, and in the second one we are checking its relation to negative 0.5. there is a period of 1 cycle in between that must be kept separate.

dalittlet replied:
Ah I missed the negative sign

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Devices/ModifiedEncoders.java:
```diff
@@ -0,0 +1,144 @@
+    }
+    
+    public double getRate(){
+        if(encoder != null){
```
Ternary

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Devices/ModifiedEncoders.java:
```diff
@@ -0,0 +1,144 @@
+    }
+
+    public double getRelativeDistance(){
+        if (encoder != null){
```
Ternary

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Devices/ModifiedEncoders.java:
```diff
@@ -0,0 +1,144 @@
+    }
+
+    public double getAbsoluteDistance(){
+        if(dutyCycleEncoder != null)
```
Ternary

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Devices/ModifiedEncoders.java:
```diff
@@ -0,0 +1,144 @@
+    }
+
+    public double getAbsolutePosition(){
+        if(dutyCycleEncoder != null){
```
Ternary

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Devices/NEOSparkMaxMotor.java:
```diff
@@ -22,6 +31,26 @@ public NEOSparkMaxMotor(int CANID){
+        motorConfig = new SparkMaxConfig();
+        closedLoopController = m_motor.getClosedLoopController();
+
+        motorConfig.encoder
```
Do these apply to all to like all Neo Spark Maxes?

dalittlet replied:
39-51

echo2334 replied:
i must talk with aiden before resolving this.

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Devices/NEOSparkMaxMotor.java:
```diff
@@ -34,25 +63,40 @@ public boolean isOperational(){
-        }*/
-        System.out.println("Speed is being set");
+        }
+        //System.out.println("Speed is being set");
```
```suggestion
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Devices/NEOSparkMaxMotor.java:
```diff
@@ -34,25 +63,40 @@ public boolean isOperational(){
         else {
             m_motor.set(speed);
-            System.out.println("Speed is " + speed);
+           // System.out.println("Speed is " + speed);
```
```suggestion
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Robot.java:
```diff
@@ -4,40 +4,68 @@
 
-  private static final String kDefaultAuto = "Default";
-  private static final String kCustomAuto = "My Auto";
+  // private static final String kDefaultAuto = "Default";
```
```suggestion
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Robot.java:
```diff
@@ -4,40 +4,68 @@
-  private static final String kDefaultAuto = "Default";
-  private static final String kCustomAuto = "My Auto";
+  // private static final String kDefaultAuto = "Default";
+  // private static final String kCustomAuto = "My Auto";
```
```suggestion
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Robot.java:
```diff
@@ -4,40 +4,68 @@
 
+    frontLimelight = frontLimelight.getInstance();
+
+    scoringControl =ScoringControl.getInstance();
```
```suggestion
    scoringControl = ScoringControl.getInstance();
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Robot.java:
```diff
@@ -63,25 +92,25 @@ public void robotPeriodic() {
+
     m_autoSelected = m_chooser.getSelected();
-    // m_autoSelected = SmartDashboard.getString("Auto Selector", kDefaultAuto);
     System.out.println("Auto selected: " + m_autoSelected);
```
```suggestion
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Subsystem/Elevator.java:
```diff
@@ -0,0 +1,128 @@
+    public static final double kElevatorkA = 0.06; // volt per acceleration 
+
+    public static final double kElevatorDrumRadius = Units.inchesToMeters(1.0); // change this to actual
+    public static final double kElevatorEncoderDistPerPulse = 2.0 * Math.PI * kElevatorDrumRadius / 4096; // place holder for encoder. this is for the sim
```
Is this comment still relevant?

echo2334 replied:
no

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Subsystem/Elevator.java:
```diff
@@ -0,0 +1,128 @@
+
+    private double elevatorOne;
+    private double elevatorTwo;
+    public double elevatorSpeed = .15; //adjust this
```
Still needs adjusted? If not, remove

echo2334 replied:
we don't need that at all

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Subsystem/Elevator.java:
```diff
@@ -0,0 +1,128 @@
+        elevatorMotorTwo.setVoltage((feedforwardOutput + pidOutput)*-1);
+        SmartDashboard.putNumber("Elevator/motorOneVoltage", feedforwardOutput+pidOutput);
+        SmartDashboard.putNumber("Elevator/goal", goal);
+        //elevatorOne = elevatorSpeed;
```
Remove commented out code

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Subsystem/Elevator.java:
```diff
@@ -0,0 +1,128 @@
+        m_mech2dRoot = m_mech2d.getRoot("Elevator Root", 10, 0);
+        m_elevatorMech2d = new MechanismLigament2d("Elevator", -1* encoder.get(), 90);
+        SmartDashboard.putData("Elevator Sim", m_mech2d);
+        //m_elevatorSim.setInput(m_motor.getSpeed()* RobotController.getBatteryVoltage());
```
```suggestion
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Subsystem/FrontLimelight.java:
```diff
@@ -0,0 +1,50 @@
+    };
+
+    public void log(){
+        logtoSmartDashboard();
```
Keep the consistent casing?
```suggestion
        logToSmartDashboard();
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Subsystem/FrontLimelight.java:
```diff
@@ -0,0 +1,50 @@
+        logtoSmartDashboard();
+    };
+
+    public boolean isEnabled(){
```
Is there a reason to have a function for isEnabled if it's always true?

echo2334 replied:
there was on the old repo, and we might re-add that feature in. it will allow us to disable things for testing easily.

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Subsystem/Mecanum.java:
```diff
@@ -1,26 +1,43 @@
+
+    private PIDController speedControlPID = new PIDController(5.5, 0, 0.00001);
+
+    private double MAX_SPEED_CONSTANT_FORWARD = 20; //TO DO: calculate this (meters per sec)
```
Has this been calcuated?

echo2334 replied:
nope. we are guessing right now.

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Subsystem/Mecanum.java:
```diff
@@ -1,26 +1,43 @@
+    private PIDController speedControlPID = new PIDController(5.5, 0, 0.00001);
+
+    private double MAX_SPEED_CONSTANT_FORWARD = 20; //TO DO: calculate this (meters per sec)
+    private double MAX_SPEED_CONSTANT_STRAFE = 20; //Meters per sec. Should calculate this too.
```
Also calculate?

echo2334 replied:
yeah, same here. just guessing for now.

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Subsystem/Mecanum.java:
```diff
@@ -40,8 +57,12 @@ public static Mecanum getInstance() {
     }
 
+    /*
+     * currently not functional
```
What does this mean?

dalittlet replied:
Why is this in if it's not functionality

echo2334 replied:
deleted in favor of the more advanced one

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Subsystem/Mecanum.java:
```diff
@@ -95,5 +153,10 @@ public boolean isEnabled() {
     }
-
+    public boolean moving() {
+        if(frontRight >= PortMap.MOVING_THRESHOLD || frontLeft >= PortMap.MOVING_THRESHOLD || rearLeft >= PortMap.MOVING_THRESHOLD || rearRight >= PortMap.MOVING_THRESHOLD) {
```
```suggestion
        return frontRight >= PortMap.MOVING_THRESHOLD || frontLeft >= PortMap.MOVING_THRESHOLD || rearLeft >= PortMap.MOVING_THRESHOLD || rearRight >= PortMap.MOVING_THRESHOLD;
```

And remove code below

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Subsystem/ScoringControl.java:
```diff
@@ -0,0 +1,183 @@
+                   elevator.reachGoal(EncoderValues.ELEVATOR_L4);
+                break;
+            
+            case "ejecting coral":
```
Should this case do anything if it doesn't actually work?

echo2334 replied:
we don't need it

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Subsystem/ScoringControl.java:
```diff
@@ -0,0 +1,183 @@
+    }
+
+    public void intakeCoral(){
+        state = "operator wants coral";
```
You might consider making your states stored constants somewhere so if you want to change it in one place it changes everywhere.

echo2334 replied:
ugh, fine.

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Teleop.java:
```diff
@@ -1,62 +1,112 @@
+    public boolean OperatorWants = false;
+    public boolean ElevatorIsUp;
+    public double EffectorSpeed = -0.2;
+    public double factorOfReduction;
```
What does this variable do? Why?

echo2334 replied:
factor of reduction is how we dynamically control the speed of the drive base when the elevator is at different heights.

dalittlet reviewed (commented):
ROASTED

echo2334 committed:
effector and elevator actions work!

- make a reset for elevator so that when disabled
- effector action cannot be used if not at a stage preset

FRC8334Students committed:
started on pull request fixes

FRC8334Students committed:
reverted failed changes to counter a stick drift problem that was solved by switching controllers. made changes to factorOfReduction to allow strafe and more freedom of movement when elevator is up

echo2334 committed:
some clean up

echo2334 committed:
L3 auto and an auto docs added

FRC8334Students committed:
most, but not all, of talyor's pull request changes have been implemented

echo2334 committed:
adding L3 to smartdashboard options

echo2334 committed:
merged rhea's current (at time of merge) best auto code with the almost ready to merge into main development code. changes were made to make them work together.

echo2334 committed:
deleted a print statement in EffectorAction.

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Actions/DriveForDistanceAction.java:
```diff
@@ -0,0 +1,64 @@
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+
+import frc.robot.Subsystem.Mecanum;
+//import frc.robot.Subsystem.PositioinEstimation;
```
```suggestion
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Actions/DriveForDistanceAction.java:
```diff
@@ -0,0 +1,64 @@
+//import frc.robot.Subsystem.PositioinEstimation;
+
+public class DriveForDistanceAction implements Actions{
+        Timer timer;
```
Mis-align and might want public or private

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Actions/DriveForDistanceAction.java:
```diff
@@ -0,0 +1,64 @@
+    private double seconds;
+
+    private Mecanum mecanum;
+    //private PositionEstimation position;
```
Remove all commented out things?

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Actions/DriveForDistanceAction.java:
```diff
@@ -0,0 +1,64 @@
+    public DriveForDistanceAction(double distance, double seconds)
+    {
+        mecanum = Mecanum.getInstance();
+        //position = PositionEstimation.getInstance();
```
```suggestion
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Actions/DriveForDistanceAction.java:
```diff
@@ -0,0 +1,64 @@
+    {
+        timer = new Timer();
+        timer.start();
+        //currentDistance = position.getDistance();
```
```suggestion
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Actions/DriveForDistanceAction.java:
```diff
@@ -0,0 +1,64 @@
+    @Override
+    public void update()
+    {
+        //currentDistance = position.getDistance();
```
```suggestion
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Actions/DriveForTimeAction.java:
```diff
@@ -0,0 +1,51 @@
+    public void start() {
+        timer = new Timer();
+        timer.start();
+        System.out.println("Current Action DriveForTimeAction Started");
```
```suggestion
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Actions/EffectorAction.java:
```diff
@@ -0,0 +1,47 @@
+    
+    private double seconds;
+    private double speed;
+    Timer timer;
```
public or private?

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Actions/MoveElevatorAction.java:
```diff
@@ -0,0 +1,43 @@
+    
+    private double seconds;
+    private ScoringControl scoringControl = null;
+    Timer timer;
```
public or private?

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Actions/TurnDegreesAction.java:
```diff
@@ -0,0 +1,72 @@
+        turn = (targetDegrees - currentDegrees)/ 180; //the power of the turn, divided to make power less
+        
+        mDrive.driveWithSpeed(0, 0, turn); // no forward, no strafe, only rotation
+        System.out.println("gyro yaw:" + gyro.getAngleDegrees());
```
```suggestion
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/AutoMissionChooser.java:
```diff
@@ -0,0 +1,162 @@
+package frc.robot.Auto;
+
+import frc.robot.Auto.Missions.*;
```
Fun fact, this .* imports everything so you don't need the lines below this

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Missions/BlueMissions/BlueScoreL2.java:
```diff
@@ -0,0 +1,32 @@
+       
+       // runAction(new WaitAction(AutoMissionChooser.delay)); // MAY NOT BE NEEDED
+        runAction(new DriveForTimeAction(-0.5,  4));
+        //runAction(new TurnDegreesAction(65, 1));// add 5 degrees because michalangelo has consistently com up 5 degrees short of the needed turn
```
```suggestion
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Missions/BlueMissions/BlueScoreL2.java:
```diff
@@ -0,0 +1,32 @@
+    @Override
+    protected void routine() throws AutoMissionEndedException {
+       
+       // runAction(new WaitAction(AutoMissionChooser.delay)); // MAY NOT BE NEEDED
```
```suggestion
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Missions/BlueMissions/BlueScoreL3.java:
```diff
@@ -0,0 +1,31 @@
+    @Override
+    protected void routine() throws AutoMissionEndedException {
+       
+       // runAction(new WaitAction(AutoMissionChooser.delay)); // MAY NOT BE NEEDED
```
```suggestion
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Missions/BlueMissions/BlueScoreL3.java:
```diff
@@ -0,0 +1,31 @@
+       
+       // runAction(new WaitAction(AutoMissionChooser.delay)); // MAY NOT BE NEEDED
+        runAction(new DriveForTimeAction(-0.5,  4));
+        //runAction(new TurnDegreesAction(65, 1));// add 5 degrees because michalangelo has consistently com up 5 degrees short of the needed turn
```
```suggestion
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Missions/BlueMissions/BlueScoreL4.java:
```diff
@@ -0,0 +1,32 @@
+    @Override
+    protected void routine() throws AutoMissionEndedException {
+       
+        //runAction(new WaitAction(AutoMissionChooser.delay)); // MAY NOT BE NEEDED
```
```suggestion
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Missions/BlueMissions/BlueScoreL4.java:
```diff
@@ -0,0 +1,32 @@
+    protected void routine() throws AutoMissionEndedException {
+       
+        //runAction(new WaitAction(AutoMissionChooser.delay)); // MAY NOT BE NEEDED
+        //runAction(new DriveForTimeAction(-0.5, 5));
```
```suggestion
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Missions/BlueMissions/BlueScoreL4.java:
```diff
@@ -0,0 +1,32 @@
+       
+        //runAction(new WaitAction(AutoMissionChooser.delay)); // MAY NOT BE NEEDED
+        //runAction(new DriveForTimeAction(-0.5, 5));
+        //runAction(new TurnDegreesAction(30, 4));// add 5 degrees because michalangelo has consistently com up 5 degrees short of the needed turn
```
```suggestion
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Missions/RedMissions/RedScoreL2.java:
```diff
@@ -0,0 +1,35 @@
+    @Override
+    protected void routine() throws AutoMissionEndedException {
+       
+        //runAction(new WaitAction(AutoMissionChooser.delay)); // MAY NOT BE NEEDED
```
```suggestion
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Missions/RedMissions/RedScoreL2.java:
```diff
@@ -0,0 +1,35 @@
+    protected void routine() throws AutoMissionEndedException {
+       
+        //runAction(new WaitAction(AutoMissionChooser.delay)); // MAY NOT BE NEEDED
+        //runAction(new DriveForTimeAction(-0.5,  4));
```
```suggestion
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Missions/RedMissions/RedScoreL2.java:
```diff
@@ -0,0 +1,35 @@
+       
+        //runAction(new WaitAction(AutoMissionChooser.delay)); // MAY NOT BE NEEDED
+        //runAction(new DriveForTimeAction(-0.5,  4));
+        //runAction(new TurnDegreesAction(65, 1));// add 5 degrees because michalangelo has consistently com up 5 degrees short of the needed turn
```
```suggestion
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Missions/RedMissions/RedScoreL3.java:
```diff
@@ -0,0 +1,33 @@
+    @Override
+    protected void routine() throws AutoMissionEndedException {
+       
+        //runAction(new WaitAction(AutoMissionChooser.delay)); // MAY NOT BE NEEDED
```
```suggestion
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Missions/RedMissions/RedScoreL3.java:
```diff
@@ -0,0 +1,33 @@
+       
+        //runAction(new WaitAction(AutoMissionChooser.delay)); // MAY NOT BE NEEDED
+        runAction(new DriveForTimeAction(-0.5,  4));
+        //runAction(new TurnDegreesAction(65, 1));// add 5 degrees because michalangelo has consistently com up 5 degrees short of the needed turn
```
```suggestion
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Missions/RedMissions/RedScoreL4.java:
```diff
@@ -0,0 +1,35 @@
+    @Override
+    protected void routine() throws AutoMissionEndedException {
+       
+        //runAction(new WaitAction(AutoMissionChooser.delay)); // MAY NOT BE NEEDED
```
```suggestion
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Missions/RedMissions/RedScoreL4.java:
```diff
@@ -0,0 +1,35 @@
+    protected void routine() throws AutoMissionEndedException {
+       
+        //runAction(new WaitAction(AutoMissionChooser.delay)); // MAY NOT BE NEEDED
+        //runAction(new DriveForTimeAction(-0.5,  4));
```
```suggestion
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Missions/RedMissions/RedScoreL4.java:
```diff
@@ -0,0 +1,35 @@
+       
+        //runAction(new WaitAction(AutoMissionChooser.delay)); // MAY NOT BE NEEDED
+        //runAction(new DriveForTimeAction(-0.5,  4));
+        //runAction(new TurnDegreesAction(30, 1));// add 5 degrees because michalangelo has consistently com up 5 degrees short of the needed turn
```
```suggestion
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Auto/Missions/RedMissions/RedScoreL4.java:
```diff
@@ -0,0 +1,35 @@
+        // String options: passive, ramp, Score L1, Score L2, Score L3, Score L4, ejecting coral
+        runAction(new MoveElevatorAction(2, state.SCOREL4)); // May be needed depending on how robot starts
+        runAction(new EffectorAction(0.12,1));
+        //runAction(new DriveForTimeAction(0.5, 0.2));
```
```suggestion
```

dalittlet commented on 2025REDOBuildSeason/src/main/java/frc/robot/Robot.java:
```diff
@@ -63,25 +95,24 @@ public void robotPeriodic() {
    */
-  @Override
+  
+   @Override
```
```suggestion
  @Override
```

echo2334 committed:
gets rid of needless comments
